tests/test_handlers_llm_provider.py

```diff
@@ -0,0 +1,379 @@
+            assert 1 <= result <= 100
+
+    @requires_openai
+    def test_with_config_params(self):
+        """Test LiteLLMProvider accepts and uses additional configuration parameters."""
+        # Test with temperature parameter
+        with (
+            handler(LiteLLMProvider(model_name="gpt-4o-mini", temperature=0.1)),
+            handler(LimitLLMCallsHandler(max_calls=1)),
+        ):
+            result = simple_prompt("deterministic test")
+            assert isinstance(result, str)
+
+
+class TestLLMLoggingHandler:
+    """Tests for LLMLoggingHandler functionality."""
+
+    @requires_openai
+    def test_logs_requests(self, caplog):
+        """Test that LLMLoggingHandler properly logs LLM requests."""
+        with caplog.at_level(logging.INFO):
+            with (
+                handler(LLMLoggingHandler()),
+                handler(LiteLLMProvider(model_name="gpt-4o-mini")),
+                handler(LimitLLMCallsHandler(max_calls=1)),
+            ):
+                result = simple_prompt("testing")
+                assert isinstance(result, str)
+
+        # Check that logging occurred
+        assert any("llm.request" in record.message for record in caplog.records)
+
+    @requires_openai
+    def test_custom_logger(self, caplog):
+        """Test LLMLoggingHandler with a custom logger."""
+        custom_logger = logging.getLogger("test_custom_logger")
+
+        with caplog.at_level(logging.INFO, logger="test_custom_logger"):
+            with (
+                handler(LLMLoggingHandler(logger=custom_logger)),
+                handler(LiteLLMProvider(model_name="gpt-4o-mini")),
+                handler(LimitLLMCallsHandler(max_calls=1)),
+            ):
+                result = simple_prompt("testing")
+                assert isinstance(result, str)
+
+        # Verify custom logger was used
+        assert any(
+            record.name == "test_custom_logger" and "llm.request" in record.message
+            for record in caplog.records
+        )
+
+
+class TestProgramSynthesis:
+    """Tests for ProgramSynthesis handler functionality."""
+
+    @requires_openai
+    def test_generates_callable(self):
+        """Test ProgramSynthesis handler generates executable code."""
+        for i in range(3):
+            try:
+                with (
+                    handler(LiteLLMProvider(model_name="gpt-4o-mini")),
+                    handler(ProgramSynthesis()),
+                    handler(LimitLLMCallsHandler(max_calls=1)),
+                ):
+                    count_func = create_function("a")
+
+                    assert callable(count_func)
+                    # Test the generated function
+                    assert count_func("banana") == 3
+                    assert count_func("cherry") == 0
+                    assert count_func("aardvark") == 3
+            except SynthesisError as e:
+                if i < 2:
+                    continue
+                raise e
```

Comment:
Let's extract the retry logic into a decorator. We'll probably want something similar for other tests.

Reply:
Done!